### PR TITLE
C#: Split `DependencyManager` into multiple classes

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR is splitting the dependency manager into multiple classes:
- file lookup, such as getting all source files, solutions, ..., are moved to `FileProvider`, and
- nuget package restoration is moved to `NugetPackageRestorer`.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs [hunk 1]
```diff
@@ -121,13 +112,14 @@ void exitCallback(int ret, string msg, bool silent)
                 throw;
             }
 
-            RestoreNugetPackages(allNonBinaryFiles, allProjects, allSolutions, dllLocations);
+            var dllLocations = fileProvider.Dlls.Select(x => new AssemblyLookupLocation(x)).ToHashSet();
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:121]: Are we sure that none of the preceding code adds any dlls to the sourcedir?

Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:121]: That's a good question. In the second commit I made the `FileProvider.all` not lazy. So the question is if there's any code that adds DLLs before the `new FileProvider(...)` call. And I think the answer is no.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs [hunk 2]
```diff
@@ -723,7 +665,7 @@ private void AnalyseProject(FileInfo project)
             }
         }
 
-        public void Dispose(TemporaryDirectory? dir, string name)
+        public static void DisposeTempDirectory(TemporaryDirectory? dir, string name, ILogger logger)
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:668]: nit: Maybe consider to put this method in another (static class) or incorporate the try/catch logic directly in the `Dispose` method in the `TemporaryDirectory` class.

**Review by michaelnebel — DISMISSED** [review]
LGTM!

**Review by michaelnebel — APPROVED** [review]
LGTM!

**Pull request merged**